simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,458 @@
+from __future__ import annotations
+
+import warnings
+
+import numpy as np
+from dataclasses import dataclass
+
+from .directives import InversionDirective, UpdatePreconditioner, BetaSchedule
+from simpeg.regularization import Sparse, BaseSparse, SmoothnessFirstOrder
+from simpeg.utils import validate_integer, validate_float
+
+
+@dataclass
+class IRLSMetrics:
+    """
+    Data class to store metrics used by the IRLS algorithm.
+
+    Parameters
+    ----------
+    input_norms : list of floats
+        List of norms temporarily stored during the initialization.
+    irls_iteration_count : int
+        Number of IRLS iterations.
+    start_irls_iter : int or None
+        Iteration number when the IRLS process started.
+    f_old : float
+        Previous value of the regularization function.
+    """
+
+    input_norms: list[float] = (2.0, 2.0, 2.0, 2.0)
+    irls_iteration_count: int = 0
+    start_irls_iter: int | None = None
+    f_old: float = 0.0
+
+
+class UpdateIRLS(BetaSchedule):
+    """
+    Directive to control the IRLS iterations for regularization.Sparse.
+
+    Parameters
+    ----------
+    cooling_rate: int
+        Number of iterations to cool beta.
+    cooling_factor: float
+        Factor to cool beta.
+    chifact_start: float
+        Starting chi factor for the IRLS iterations.
+    chifact_target: float
+        Target chi factor for the IRLS iterations.
+    irls_cooling_factor: float
+        Factor to cool the IRLS threshold epsilon.
+    f_min_change: float
+        Minimum change in the regularization function to continue the IRLS iterations.
+    max_irls_iterations: int
+        Maximum number of IRLS iterations.
+    misfit_tolerance: float
+        Tolerance for the target misfit.
+    percentile: int
```

Comment:
Small change on the type of the `percentile` argument:
```suggestion
    percentile: float
```